[AIRFLOW-5636] Allow adding or overriding existing Operator Links for…

kaxil opened the pull request:
… an Operator

This will be done via Airflow Plugins

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5636


### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Currently, we can add Operator link globally for all tasks or add operator links in existing tasks.

We should be able to add Operator links on a Single Operator or override existing Operator link on an operator. 

This PR adds this functionality. This would also help in adding Operator links once we enable Serialized DAGs #5743  

In the future, this will become a **de facto** way of adding OperatorLinks.


### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

kaxil committed:
[AIRFLOW-5636] Allow adding or overriding existing Operator Links for an Operator

This will be done via Airflow Plugins

kaxil committed:
Update baseoperator.py

milton0825 commented on airflow/models/baseoperator.py:
```diff
@@ -566,7 +566,20 @@ def priority_weight_total(self):
+
+        op_extra_links_from_plugin = {}
+        for ope in operator_extra_links:
+            if ope.operator_name == self.__class__.__name__:
```
Shall we also include the module name here? It is possible that the user create an operator with the same name but registered in a different package through plugin?

kaxil replied:
Updated the PR

milton0825 commented on airflow/models/baseoperator.py:
```diff
@@ -1057,6 +1070,17 @@ def name(self) -> str:
+        This property will be used by Airflow Plugins to find the class to which you want
+        to assign this Operator Link
+
+        :return: Name of Operator used by task for which
```
Probably the full class path here?

ashb replied:
Yeah, I think full class path, or perhaps the class itself?

kaxil replied:
Updated the PR

ashb replied:
Oh, docs still need updating though.

ashb replied:
Does this need to be a property? Could we instead just do

```
    operators = None  # type: Optional[List[BaseOperator]]
    """
    This property will be used by Airflow Plugins to find the Operators to which you want
    to assign this Operator Link
    """
```

(It's slightly odd, but to document a property the string goes after it.

kaxil replied:
Updated

kaxil committed:
Allow Extra Operator Links to have operator classes with same name

ashb commented on tests/www/test_views.py:
```diff
@@ -1854,6 +1856,24 @@ def test_extra_links_no_response(self, get_dag_function):
+        if isinstance(response.data, bytes):
+            response_str = response_str.decode()
+        self.assertEqual(json.loads(response_str), {
+            'url': 'https://airflow.apache.org/1.10.5/',
```
I'm not quite sure what this is testing, nor where this string is coming from in the code?

kaxil replied:
It is coming from here:
https://github.com/apache/airflow/blob/d93bb34ca08d45d4dce77bc0e6415b4251c53b3d/tests/plugins/test_plugin.py#L113-L118

It is testing that if OperatorLink is added in the plugin, the same link appears

kaxil replied:
Let me move them to separate file

ashb replied:
A comment explaining what is going on here would help I think (that the link is being overridden by one registered via a plugin)

kaxil replied:
Added the comment

kaxil committed:
Allow operator links to be added to multiple operators

ashb commented on tests/plugins/test_plugin.py:
```diff
@@ -23,10 +23,11 @@
 # This is the class you derive to create a plugin
 from airflow.plugins_manager import AirflowPlugin
 from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from tests.utils.extra_link_operators import AirflowLink, AirflowLink2, GithubLink, GoogleLink
```
I'm not 100% happy with this module name - not sure I have a better one, but these aren't "utils"

kaxil replied:
Changed

kaxil committed:
Changes after review

kaxil committed:
test commit to see if it fixes

kaxil committed:
Revert "test commit to see if it fixes"

This reverts commit 895da8bccd9d85773e9c71b7b2ccc22b22e6f96b.

kaxil committed:
Fix tests

Found out that the issue was due to imp.load_source as it can not import dependencies

https://stackoverflow.com/q/9066777/5691525

kaxil committed:
Move Helper Operators & Extra Links to airflow/utils/tests.py

This are moved because they are needed by both plugin_manager.py (as it needs to import using imp.load_source when loading test_plugin.py) and  test_views.py.

kaxil committed:
Fix typo in docs

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -1048,6 +1061,14 @@ class BaseOperatorLink(metaclass=ABCMeta):
     Abstract base class that defines how we get an operator link.
     """
 
+    operators = None   # type: Optional[List[BaseOperator]]
```
It might make the code a bit cleaner if we do this:

```suggestion
    operators = []   # type: List[BaseOperator]
```

kaxil replied:
This was done mainly for mutable argument gotcha. 

https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

ashb replied:
Not an argument though.

kaxil replied:
Jeez !! yeah

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -1048,6 +1061,14 @@ class BaseOperatorLink(metaclass=ABCMeta):
     Abstract base class that defines how we get an operator link.
     """
 
+    operators = None   # type: Optional[List[BaseOperator]]
```
```suggestion
    operators = []   # type: List[Type[BaseOperator]]
```

Oh also we want a list of classes, not a list of objects!

kaxil replied:
Updated

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -1048,6 +1061,14 @@ class BaseOperatorLink(metaclass=ABCMeta):
+    This property will be used by Airflow Plugins to find the Operators to which you want
+    to assign this Operator Link
+
+    :return: List of Operator used by task for which you want to create extra link
```
```suggestion
    :return: List of Operator classes used by task for which you want to create extra link
```

kaxil replied:
Updated

ashb commented on airflow/plugins_manager.py:
```diff
@@ -181,6 +194,7 @@ def make_module(name, objects):
 flask_appbuilder_menu_links = []  # type: List[Any]
 stat_name_handler = None  # type: Any
 global_operator_extra_links = []  # type: List[Any]
+operator_extra_links = []  # type: List[Any]
```
```suggestion
operator_extra_links = []  # type: List[BaseOperatorLink]
```

I think?

kaxil replied:
Updated

ashb commented on airflow/plugins_manager.py:
```diff
@@ -205,6 +219,10 @@ def make_module(name, objects):
+    # Only register Operator links if its ``operators`` property is not None
+    # So that we can only attach this links to a specific Operator
+    operator_extra_links.extend([
+        ope for ope in p.operator_extra_links if ope.operators is not None])
```
`if ope.operators` if we change it to not be Optional

kaxil replied:
Updated

ashb commented on docs/howto/define_extra_link.rst:
```diff
@@ -56,3 +56,75 @@ The following code shows how to add extra links to an operator:
+---------------------------------------
+
+You can also add (or override) an extra link to an existing operators
+through Airflow plugin.
```
```suggestion
through an Airflow plugin.
```

kaxil replied:
Updated

ashb commented on docs/howto/define_extra_link.rst:
```diff
@@ -56,3 +56,75 @@ The following code shows how to add extra links to an operator:
+For example, the following Airflow plugin will add an Operator Link on all
+tasks using :class:`~airflow.operators.gcs_to_s3.GoogleCloudStorageToS3Operator` operator.
+
+**Adding Operator Links to Existing Operators**
```
~I don't think we want this line do we?~ Oh I see.

ashb commented on docs/howto/define_extra_link.rst:
```diff
@@ -56,3 +56,75 @@ The following code shows how to add extra links to an operator:
 :ref:`plugin example <plugin-example>`.
+
+
+Add Operator Link to Existing Operators
```
```suggestion
Add or override Links to Existing Operators
```

kaxil replied:
Updated

ashb commented on docs/howto/define_extra_link.rst:
```diff
@@ -56,3 +56,75 @@ The following code shows how to add extra links to an operator:
+
+**Overriding Operator Links of Existing Operators**:
+
+For :class:`~airflow.gcp.operators.bigquery.BigQueryOperator` you can create
```
This needs re-wording slightly, as one reading might take this to mean it only applies to this one class.

ashb replied:
```
It is also possible to replace a built in link on an operator via a Plugin. For example 
:class:`~airflow.gcp.operators.bigquery.BigQueryOperator` includes a link to the GCP
console, but if we wanted to change that link we could:

kaxil replied:
Updated

kaxil committed:
Updates after PR review

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -566,7 +566,20 @@ def priority_weight_total(self):
+        from airflow.plugins_manager import operator_extra_links
+
+        op_extra_links_from_plugin = {}
+        for ope in operator_extra_links:
```
```suggestion
        for ope in self.operator_extra_links:
```

If this doesn't fail tests our tests need improving!

kaxil replied:
no, it is intentional.

`operator_extra_links` is from plugins

`self.operator_extra_links` is from the operator itself

kaxil replied:
Check line 569

ashb replied:
Temporary blindness!

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -566,7 +566,20 @@ def priority_weight_total(self):
+
+        op_extra_links_from_plugin = {}
+        for ope in operator_extra_links:
+            if ope.operators and self.__class__ in ope.operators:
```
```suggestion
            if type(self) in ope.operators:
```

Is enough now I think? (and switch to `type(self)` as it avoids accessing `__` vars.

kaxil replied:
Agree

kaxil replied:
Actually pylint complains with:

```
Using type() instead of isinstance() for a typecheck.
```

ashb replied:
Bah, stupid pylint.
